presto-main/src/main/java/com/facebook/presto/operator/aggregation/PrecisionRecallAggregation.java

```diff
@@ -0,0 +1,175 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation;
+
+import com.facebook.presto.operator.aggregation.state.PrecisionRecallState;
+import com.facebook.presto.spi.block.BlockBuilder;
+import com.facebook.presto.spi.function.AggregationFunction;
+import com.facebook.presto.spi.function.AggregationState;
+import com.facebook.presto.spi.function.CombineFunction;
+import com.facebook.presto.spi.function.Description;
+import com.facebook.presto.spi.function.InputFunction;
+import com.facebook.presto.spi.function.OutputFunction;
+import com.facebook.presto.spi.function.SqlType;
+import com.facebook.presto.spi.type.DoubleType;
+import com.facebook.presto.spi.type.StandardTypes;
+
+import java.util.Iterator;
+import java.util.Map;
+import java.util.SortedMap;
+import java.util.TreeMap;
+
+@AggregationFunction("precision_recall")
+@Description("Computes precision-recall curves")
+public final class PrecisionRecallAggregation
+{
+    @InputFunction
+    public static void input(
+            @AggregationState PrecisionRecallState state,
+            @SqlType(StandardTypes.BIGINT) long numBins,
+            @SqlType(StandardTypes.BOOLEAN) boolean outcome,
+            @SqlType(StandardTypes.DOUBLE) double pred,
+            @SqlType(StandardTypes.DOUBLE) double weight)
+    {
+        if (state.getTrueWeights() == null) {
+            state.setTrueWeights(new FixedNumericHistogram((int) (numBins), 0.0, 1.0));
+            state.setFalseWeights(new FixedNumericHistogram((int) (numBins), 0.0, 1.0));
+        }
+
+        if (pred < 0 || pred > 1 || weight < 0.0) {
```

Comment:
Couldn't this be performed outside of the function itself?  Perhaps someone would prefer it to throw rather than treat it as null?  But if the default is to throw, one could always perform this transformation outside of the function, however the converse isn't true.

Reply:
Not sure I get it. It is a violation of the contract of the function, in some sense, just like changing the number of bins between instances (which should be constant). IMHO, these should all be treated the same way. Are you suggesting to use `checkArgument`, or just not check (and whatever will happen is on the user), or return `NULL`?

Reply:
We can throw a PrestoException which indicates a data quality issue.  i.e. we can throw `new PrestionException(INVALID_FUNCTION_ARGUMENT, ...)` if these things occur.  If you don't care about that thought, you can wrap the function in SQL with `TRY`, which will return null if that gets thrown.